core/shared/src/main/scala/scalaz/zio/Promise.scala

```diff
@@ -129,7 +129,7 @@ class Promise[E, A] private (private val state: AtomicReference[State[E, A]]) ex
       retry = !state.compareAndSet(oldState, newState)
     }
 
-    if (joiners ne null) joiners.reverse.foreach(k => executionContext.execute(() => k(r)))
+    if (joiners ne null) joiners.reverse.foreach(k => env.executor(Executor.Type.Asynchronous).submit(() => k(r)))
```

Comment:
maybe some task could be run with `scalaz.zio.internal.Executor.Type.Synchronous`?
Then the `Callback` should be `AsyncCallback` and `SyncCallback`.

Reply:
This code is broken currently (i.e. not respecting executor preferences) and will be fixed shortly.